src/metabase/api/api_key.clj

```diff
@@ -0,0 +1,75 @@
+(ns metabase.api.api-key
+  "/api/api-key endpoints for CRUD management of API Keys"
+  (:require
+   [compojure.core :refer [POST]]
+   [crypto.random :as crypto-random]
+   [metabase.api.common :as api]
+   [metabase.models.api-key :as api-key]
+   [metabase.models.permissions-group :as perms-group]
+   [metabase.models.user :as user]
+   [metabase.util :as u]
+   [metabase.util.honey-sql-2 :as h2x]
+   [metabase.util.malli.schema :as ms]
+   [toucan2.core :as t2]))
+
+
+(def ^:private user-first-name "API Key:")
+
+(defn- full-name [n]
+  (str user-first-name " " n))
+
+(defn- count-api-key-users-with-name [n]
+  (:count
+   (t2/query-one {:select [[[:count :core_user.id] :count]]
+                  :from :core_user
+                  :where [:= (h2x/concat :first_name (h2x/literal " ") :last_name) (full-name n)]})))
+
+(defn- key-with-unique-prefix []
+  (u/auto-retry 5
+   (let [api-key (api-key/generate-key)
+         prefix (api-key/prefix api-key)]
+     ;; we could make this more efficient by generating 5 API keys up front and doing one select to remove any
+     ;; duplicates. But a duplicate should be rare enough to just do multiple queries for now.
+     (if-not (t2/exists? :model/ApiKey :key_prefix prefix)
+       api-key
+       (throw (ex-info "could not generate key with unique prefix" {}))))))
+
+(api/defendpoint POST "/"
+  "Create a new API key (and an associated `User`) with the provided name and group ID."
+  [:as {{:keys [group_id name] :as _body} :body}]
+  {group_id ms/PositiveInt
+   name     ms/NonBlankString}
+  (let [api-key (key-with-unique-prefix)
+        prefix  (api-key/prefix api-key)
+        email   (format "api-key-user-%s@api-key.invalid" name)]
+    (api/check-superuser)
+    (api/checkp (not (t2/exists? :model/User :%lower.email email))
+      "name" "An API key with this name already exists.")
+    (t2/with-transaction [_conn]
+      (let [user (first (t2/insert-returning-instances! :model/User
+                                                        {:email      email
+                                                         :password   (crypto-random/base64 16)
+                                                         :first_name "API Key:"
+                                                         :last_name  name
```

Comment:
I just realized something that may require a change of approach here: `API Key` should almost certainly be translated, but it won't be if we're storing it as a raw string in the DB. 

Could you change it to be a i18n'ed string stored in memory that is hydrated on the `User` model when an internal user is fetched? 

Another schema adjustment that might make sense is to store the API key name in the `api_key` table and hydrate the `last_name` field when internal users are fetched, rather than storing it directly in the `core_user` table. This would solve the potential race condition below, right? Unless I'm misunderstanding something there.

i.e. an `api-key` user would just be a shell and all data related to the API key would be stored in the `api_key` table, instead of split across that and the `core_user` table

Reply:
I like this and implemented it! Thanks for the idea.